### PR TITLE
Allow inject secret hash without secret

### PR DESCRIPTION
**Pull request by offerm**
Feature: Allow inject secret-hash without a secret

With this commit the REST api allows the caller to provide a secret-hash without a secret. The intention is that the secret will be provided by the payee(target).
Since the target does not yet support this capability, in the case that only secret-hash provided, the code ignores the secret-hash, select a secret an use it. This will be removed next when the state machine is enhanced to support this mode.  
The EventPaymentsSecretSuccess includes now the secret so that the API can should it to the user. The secret and secrethash removed from payment_status.

Points to consider:
- How the secret is returned to the api level (payment_status)

### CODE REVIEW

**Comment by offerm** [comment]
@rakanalh @LefterisJP this one is ready as far as I'm concerned. Happy to get your comments.

**Comment by offerm** [comment]
@rakanalh yes, you are right.
please see what I wrote in the gitter chat:

I filed 3 PRs which are build one on top of the other. The first raiden-network/raiden#3560 is still dealing with the interface and allows payment call with secret-hash only. On top of that there is raiden-network/raiden#3577 which simulates the resolver and include that change to the state machine (event handler). Finally there is raiden-network/raiden#3578 which includes the full resolver support and some code refactoring to make the changes smaller.
raiden-network/raiden#3578 can be reviewed directly. it can also be done one by one to make it simpler.

File: raiden/raiden_service.py [hunk 1]
```diff
@@ -891,18 +891,22 @@ def start_mediated_transfer_with_secret(
                 amount=amount,
                 token_network_identifier=token_network_identifier,
                 payment_done=AsyncResult(),
-                secret=secret,
-                secret_hash=secret_hash,
             )
             self.targets_to_identifiers_to_statuses[target][identifier] = payment_status
 
+        # until we have full support create the secret here and ignore the provided hash
```
Comment (rakanalh) [inline raiden/raiden_service.py]: I know what you're referring to when you say "until we have support", but others won't. Please clarify this comment

File: raiden/transfer/mediated_transfer/state.py [hunk 2]
```diff
@@ -627,8 +627,10 @@ def __init__(
             initiator: InitiatorAddress,
             target: TargetAddress,
             secret: Secret,
+            secret_hash: SecretHash = None,
```
[the remark this answers is not in the document]
Comment (rakanalh) [reply raiden/transfer/mediated_transfer/state.py]: Also, our naming convention is to use `secrethash` instead of `secret_hash`

**Comment by LefterisJP** [comment]
@offerm Why make many small PRs that don't really stand alone? It starts to become quite hard to follow the functionality you are trying to introduce in the codebase and how they all tie together like this.

**Comment by offerm** [comment]
> @offerm Why make many small PRs that don't really stand alone? It starts to become quite hard to follow the functionality you are trying to introduce in the codebase and how they all tie together like this.

@LefterisJP Sorry about it. I created these PR with a hope that it will be easier to understand what I did. 
As I wrote in the gitter channel:
>@LefterisJP I filed 3 PRs which are build one on top of the other. The first raiden-network/raiden#3560 is still dealing with the interface and allows payment call with secret-hash only. On top of that there is raiden-network/raiden#3577 which simulates the resolver and include that change to the state machine (event handler). Finally there is raiden-network/raiden#3578 which includes the full resolver support and some code refactoring to make the changes smaller.
raiden-network/raiden#3578 can be reviewed directly. it can also be done one by one to make it simpler.

With your permission, I suggest to close #3560  and #3577 and focus exclusively on #3578 which is a complete standalone PR. Let me know if you are OK with this.

File: raiden/api/python.py [hunk 3]
```diff
@@ -759,12 +759,12 @@ def transfer_async(
...
             secret_hash = to_bytes(hexstr=secret_hash)
 
-        if secret is None and secret_hash is not None:
-            raise InvalidSecretOrSecretHash('secret_hash without a secret is not supported yet.')
-
         if secret is not None and secret_hash is not None and secret_hash != sha3(secret):
             raise InvalidSecretOrSecretHash('provided secret and secret_hash do not match.')
 
+        if secret is None and secret_hash is not None:
+            pass
```
[the remark this answers is not in the document]
Comment (offerm) [reply raiden/api/python.py]: Yes, This can be removed. However, I placed it in to make sure that any reader will understand that this is a valid option. Let me know if you like me to remove it.

File: raiden/api/rest.py [hunk 4]
```diff
@@ -1101,15 +1102,17 @@ def initiate_payment(
                 status_code=HTTPStatus.CONFLICT,
             )
 
+        secret = payment_status.payment_done.get()
```
[the remark this answers is not in the document]
Comment (offerm) [reply raiden/api/rest.py]: @rakanalh consider the case that the user provides only the secret_hash without the secret - In such case, the secret will not be stored in the payment_status and should be provided back to the api layer.

File: raiden/raiden_service.py [hunk 5]
```diff
@@ -891,18 +891,22 @@ def start_mediated_transfer_with_secret(
                 amount=amount,
                 token_network_identifier=token_network_identifier,
                 payment_done=AsyncResult(),
-                secret=secret,
-                secret_hash=secret_hash,
```
[the remark this answers is not in the document]
Comment (offerm) [reply raiden/raiden_service.py]: Again, as I wrote aboe, when the user provides only the secret_hash, the secret itself is unknown and must be provided via the AsyncResult. To make the code generic and clean, I think it is better to remove the secret and the secret_hash from here, return the secret via the AsyncResult and hash it to get the secert_hash.

File: raiden/transfer/mediated_transfer/state.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (offerm) [reply raiden/transfer/mediated_transfer/state.py]: Will fix according to naming convention.

Note that in our interesting use case, the user provides the secrethash without the secret (will be None). As such, it needs to be provided as an argument.

File: raiden/raiden_event_handler.py [hunk 6]
```diff
@@ -210,7 +210,7 @@ def handle_paymentsentsuccess(
         # With the introduction of the lock we should always get
         # here only once per identifier so payment_status should always exist
         # see: https://github.com/raiden-network/raiden/pull/3191
-        payment_status.payment_done.set(True)
+        payment_status.payment_done.set(payment_sent_success_event.secret)
```
[the remark this answers is not in the document]
Comment (offerm) [reply raiden/raiden_event_handler.py]: OK. Will try.

[the remark this answers is not in the document]
Comment (offerm) [reply raiden/raiden_event_handler.py]: Well, payment_status is a NamedTuple and can't be changed. This is why I did it as I did.

**Comment by offerm** [comment]
Will open a new PR once changes to the resolver API  implemented
